src/hotspot/share/oops/jmethodIDTable.cpp

```diff
@@ -0,0 +1,186 @@
+#include "memory/allocation.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/jmethodIDTable.hpp"
+#include "oops/method.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "utilities/concurrentHashTable.inline.hpp"
+#include "utilities/concurrentHashTableTasks.inline.hpp"
+#include "utilities/macros.hpp"
+
+// Save (jmethod, Method*) in a hashtable to lookup Method.
+// The CHT is for performance because it has lock free lookup.
+
+static uint64_t _jmethodID_counter = 0;
+
+class JmethodEntry {
+ public:
+  uint64_t _id;
+  Method*  _method;
+
+  JmethodEntry(uint64_t id, Method* method) : _id(id), _method(method) {}
+};
+
+class JmethodIDTableConfig : public AllStatic {
+ public:
+  typedef JmethodEntry Value;
+  static void* allocate_node(void* context, size_t size, Value const& value) {
+    return AllocateHeap(size, mtJNI);
+  }
+  static void free_node(void* context, void* memory, Value const& value) {
+    FreeHeap(memory);
+  }
+  static uintx get_hash(Value const& value, bool* is_dead) {
+    *is_dead = false;
+    return value._id;
+  }
+  static bool is_dead(Value const& value) {
+    return false;
+  }
+};
+
+using MethodIdTable = ConcurrentHashTable<JmethodIDTableConfig, mtJNI>;
+static MethodIdTable* _jmethod_id_table = nullptr;
+
+void JmethodIDTable::initialize() {
+  const size_t start_size = 10;
+  // 2^24 is max size
+  const size_t end_size = 24;
+  // If a chain gets to 32 something might be wrong.
+  const size_t grow_hint = 32;
+
+  _jmethod_id_table  = new MethodIdTable(start_size, end_size, grow_hint);
+}
+
+class JmethodIDLookup : StackObj {
+ private:
+  uint64_t _mid;
+
+ public:
+  JmethodIDLookup(const uint64_t mid) : _mid(mid) {}
+  uintx get_hash() const {
+    return _mid;
+  }
+  bool equals(JmethodEntry* value) {
+    return _mid == value->_id;
+  }
+
+  bool is_dead(JmethodEntry* value) {
+    return false;
+  }
+};
+
+static JmethodEntry* get_jmethod_entry(jmethodID mid) {
+  assert(mid != nullptr, "JNI method id should not be null");
+
+  Thread* current = Thread::current();
+  JmethodIDLookup lookup((uint64_t)mid);
+  JmethodEntry* result = nullptr;
+  auto get = [&] (JmethodEntry* value) {
+    // function called if value is found so is never null
+    result = value;
+  };
+  bool needs_rehashing = false;
+  _jmethod_id_table->get(current, lookup, get, &needs_rehashing);
+  assert(!needs_rehashing, "should never need rehashing");
+  return result;
+}
+
+Method* JmethodIDTable::resolve_jmethod_id(jmethodID mid) {
+  JmethodEntry* result = get_jmethod_entry(mid);
+  return result == nullptr ? nullptr : result->_method;
+}
+
+const unsigned _resize_load_trigger = 5;       // load factor that will trigger the resize
+
+static unsigned table_size(Thread* current) {
+  return 1 << _jmethod_id_table->get_size_log2(current);
+}
+
+static bool needs_resize(Thread* current) {
+  return ((_jmethodID_counter > (_resize_load_trigger * table_size(current))) &&
+         !_jmethod_id_table->is_max_size_reached());
```

Comment:
Should we not just have a separate jmethodID entry count variable we use here instead, that is incremented and decremented on insert and remove. Rather than using the next jmethodID counter which just grows monotonically regardless of and removals.

Reply:
Interesting suggestion to consider. I'm not sure yet if it is really important.

Reply:
This makes sense because we want to know how many items are in the table vs. how many items have been in the table.